x/sentry/errorreport/lambda.go

```
@@ -0,0 +1,97 @@
+)
+
+// LambdaOption is a function type that can be supplied to alter the behaviour of the
+// Lambda*Middleware functions.
+type LambdaOption func(o *lambdaOptions)
+
+// lambdaOptions configures the way Sentry is used in the context of a
+// Lambda handler wrapper.
+type lambdaOptions struct {
+	// Repanic configures whether to panic again after recovering from a panic.
+	// Use this option if you have other panic handlers or want the default
+	// behavior from AWS lambda runtime. Defaults to true.
+	Repanic bool
+}
+
+// WithRepanic configures whether to panic again after reporting an error to
+// Sentry. This setting defaults to true, as typically the function invocation
+// should be allowed to fail by the standard Lambda mechanisms once a panic
+// occurs.
+func WithRepanic(repanic bool) LambdaOption {
+	return func(o *lambdaOptions) {
+		o.Repanic = repanic
+	}
+}
+
+// LambdaMiddleware[TIn] provides error-handling middleware for a Lambda
+// function that has a payload type of TIn. This suits Lambda functions like
+// event processors, where the return has no payload.
+func LambdaMiddleware[TIn any](nextHandler lambdafunction.HandlerOf[TIn], config ...LambdaOption) lambdafunction.HandlerOf[TIn] {
+	options := configure(config)
+
+	return func(ctx context.Context, event TIn) error {
+		defer beforeHandler(ctx, options)()
+
+		err := nextHandler(ctx, event)
+
+		afterHandler(ctx, err)
+
+		return err
+	}
+}
+
+// LambdaWithOutputMiddleware[TIn, TOut] provides error-handling middleware for
+// a Lambda function that has a payload type of TIn and returns the tuple TOut,error.
+func LambdaWithOutputMiddleware[TIn any, TOut any](nextHandler lambdafunction.HandlerWithOutputOf[TIn, TOut], config ...LambdaOption) lambdafunction.HandlerWithOutputOf[TIn, TOut] {
+	options := configure(config)
+
+	return func(ctx context.Context, event TIn) (TOut, error) {
+		defer beforeHandler(ctx, options)()
+
+		out, err := nextHandler(ctx, event)
+
+		afterHandler(ctx, err)
```

Do we need separate before and after handlers, or can we report the error in the beforeHandler?

Any context mutations performed by the nextHandler won't be propagated to the afterHandler.

---

We can't handle the error in beforeHandler like this until it's returned by nextHandler() right? So if we wanna merge beforeHandler and afterHandler, does that mean we gonna end up with:
```
out, err := nextHandler(ctx, event)
defer afterHandler(ctx, options, err) // doing error reporting and panic handling here
```
Would that cause any panic being missed...?

---

Ahh yeah, I'm not sure how I missed that!

---

Marking this as resolved: the current state is correct AFAICT